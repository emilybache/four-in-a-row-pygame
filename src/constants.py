# CONSTANTS, Game Logic #
RED, YELLOW, EMPTY = range(3)
COLUMNS = 7
ROWS = 6
DROP_DELAY_MS = 250

# CONSTANTS, Graphics #
WIDTH, HEIGHT = 1100, 800
CENTER = (WIDTH // 2, HEIGHT // 2)
CENTER_X, CENTER_Y = CENTER
SCREENDIM = (WIDTH, HEIGHT)
DISC_DIAMETER = 90
DISC_RADIUS = DISC_DIAMETER // 2
BIG_TEXT = 45
MID_TEXT = 30
BOARD_WIDTH = DISC_DIAMETER * COLUMNS
BOARD_HEIGHT = DISC_DIAMETER * ROWS
BOARD_LEFT = (WIDTH - BOARD_WIDTH) // 2
BOARD_RIGHT = (WIDTH + BOARD_WIDTH) // 2
FPS = 60


class Color:
    WHITE = (255, 255, 255)
    BLACK = (0, 0, 0)
    RED = (255, 0, 0)
    GREEN = (0, 255, 0)
    BLUE = (0, 0, 255)
    YELLOW = (255, 255, 0)