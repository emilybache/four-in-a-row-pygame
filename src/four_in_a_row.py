from collections import defaultdict

# CONSTANTS #
DISC_SIZE = 80
RED, YELLOW, EMPTY = range(3)
WIDTH, HEIGHT = 1024, 768


class Color:
    WIDTH = 480
    HEIGHT = 600
    FPS = 60
    POWERUP_TIME = 5000
    BAR_LENGTH = 100
    BAR_HEIGHT = 10
    WHITE = (255, 255, 255)
    BLACK = (0, 0, 0)
    RED = (255, 0, 0)
    GREEN = (0, 255, 0)
    BLUE = (0, 0, 255)
    YELLOW = (255, 255, 0)


# STATES #

class GameState(object):
    def __init__(self):
        self.board = defaultdict(lambda: EMPTY)
        self.whos_turn_is_it = RED

    def whos_turn(self):
        return print_color(self.whos_turn_is_it)


class GameOverState(object):
    def __init__(self, winner):
        self.winner = winner


# MESSAGES #

class ColumnWasClicked(object):
    def __init__(self, column):
        self.column = column


# FUNCTIONS #

def update(model, msg):
    if isinstance(model, GameState):
        if isinstance(msg, ColumnWasClicked):
            column = msg.column
            model.board = place_brick(model.board, model.whos_turn_is_it, column)
            model.whos_turn_is_it = (model.whos_turn_is_it + 1) % 2
            won = check_winning_state(model.board, RED)
            if won:
                return GameOverState(winner=RED)

    return model


def place_brick(board, color, column):
    print(f"Placing brick color {color}")
    for i in range(6):
        y = 5 - i
        print(f"checking {column, y}")
        if board[(column, y)] == EMPTY:
            board[(column, y)] = color
            break
    return board


def print_color(color):
    return 'red' if color == RED else 'yellow'


def extract(board, pos, dir):
    cells = []
    for i in range(4):
        p = (pos[0] + dir[0] * i, pos[1] + dir[1] * i)
        cells.append(board[p])
    return cells


def check_winning_state(board, color):
    for y in range(6):
        for x in range(7):
            for dir in [(0, 1), (1, 0), (1, 1)]:
                cells = extract(board, (x, y), dir)
                if all(cell == color for cell in cells):
                    return True
    return False


def view(model, screen):
    if isinstance(model, GameState):
        screen.fill(Color.BLACK)
        for y in range(6):
            for x in range(7):
                value = model.board[(x, y)]
                if value != EMPTY:
                    color = Color.RED if value == RED else Color.YELLOW
                    x0 = int(WIDTH/2 - 7 * DISC_SIZE/2 + x * DISC_SIZE)
                    y0 = int(HEIGHT/2 - 6 * DISC_SIZE/2 + y * DISC_SIZE)
                    pos = (x0, y0)
                    pygame.draw.circle(screen, color, pos, DISC_SIZE//2, DISC_SIZE//2)

    draw_text(screen, "Press [ENTER] To Begin", 30, WIDTH / 2, HEIGHT / 2)
    draw_text(screen, "or [Q] To Quit", 30, WIDTH / 2, (HEIGHT / 2) + 40)


def draw_text(surf, text, size, x, y):
    font = pygame.font.Font(font_name, size)
    text_surface = font.render(text, True, Color.WHITE)
    text_rect = text_surface.get_rect()
    text_rect.midtop = (x, y)
    surf.blit(text_surface, text_rect)


# MAIN PROGRAM #

import pygame


def mainloop(screen):
    model = GameState()
    msgs = []
    while True:
        # Translate low level events to domain events
        ev = pygame.event.poll()
        if ev.type == pygame.KEYDOWN:
            if ev.key == pygame.K_RETURN:
                msgs.append(ColumnWasClicked(0))
            elif ev.key == pygame.K_q:
                break
        elif ev.type == pygame.QUIT:
            break

        # Handle events updating state
        while len(msgs) > 0:
            msg = msgs.pop(0)
            model = update(model, msg)

        # Display current model
        view(model, screen)

        pygame.display.update()


if __name__ == '__main__':
    pygame.init()
    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    pygame.display.set_caption("Four in a row")
    clock = pygame.time.Clock()
    font_name = pygame.font.match_font('arial')
    mainloop(screen)
    pygame.quit()
