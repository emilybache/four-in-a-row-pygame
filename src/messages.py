from collections import namedtuple

LeftMouseDownAt = namedtuple('LeftMouseDownAt', 'pos')
LeftMouseUpAt = namedtuple('LeftMouseUpAt', 'pos')
ColumnWasClicked = namedtuple('ColumnWasClicked', 'column')
MouseMovedTo = namedtuple('MouseMovedTo', 'pos')
Tick = namedtuple('Tick', 'time')