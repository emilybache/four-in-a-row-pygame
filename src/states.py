from collections import defaultdict, namedtuple

from src.constants import CENTER, RED, EMPTY


class StartScreenState(object):
    def __init__(self):
        self.time = 0
        self.music_playing = False


class GameState(object):
    def __init__(self):
        self.board = empty_board()
        self.whos_turn_is_it = RED
        self.mouse_pos = CENTER
        self.mouse_down_time = None
        self.time = None


def empty_board():
    return defaultdict(lambda: EMPTY)


GameOverState = namedtuple('GameOverState', 'winner board')
