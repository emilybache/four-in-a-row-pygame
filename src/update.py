from src.constants import DROP_DELAY_MS, RED, YELLOW, ROWS, EMPTY, COLUMNS, BOARD_LEFT, BOARD_RIGHT, DISC_DIAMETER
from src.messages import LeftMouseDownAt, Tick, ColumnWasClicked, MouseMovedTo, LeftMouseUpAt
from src.states import StartScreenState, GameState, GameOverState


def update(model, msg, audio_api):
    if isinstance(model, StartScreenState):
        if isinstance(msg, LeftMouseDownAt):
            audio_api.stop_music()
            return GameState()
        if isinstance(msg, Tick):
            if not model.music_playing:
                audio_api.play_music('music')
                model.music_playing = True
            model.time = msg.time
            return model
    if isinstance(model, GameState):
        if isinstance(msg, Tick):
            model.time = msg.time
            if model.mouse_down_time:
                if model.time > model.mouse_down_time + DROP_DELAY_MS:
                    model.mouse_down_time = None
                    model = update(model, ColumnWasClicked(convert_to_column(model.mouse_pos[0])), audio_api)
        if isinstance(msg, MouseMovedTo):
            model.mouse_pos = msg.pos
        if isinstance(msg, LeftMouseDownAt):
            if convert_to_column(msg.pos[0]) is not None:
                model.mouse_down = msg.pos
                model.mouse_down_time = model.time
        if isinstance(msg, LeftMouseUpAt):
            model.mouse_down_time = None
        if isinstance(msg, ColumnWasClicked):
            audio_api.play_sound('drop')
            model.board = place_brick(model.board, model.whos_turn_is_it, msg.column)
            model.whos_turn_is_it = (model.whos_turn_is_it + 1) % 2
            for color in [RED, YELLOW]:
                won = check_winning_state(model.board, color)
                if won:
                    return GameOverState(winner=color, board=model.board)
    if isinstance(model, GameOverState):
        if isinstance(msg, LeftMouseDownAt):
            return StartScreenState()

    return model


def place_brick(board, color, column):
    log(f"Placing brick color {print_color(color)} in column {column}")
    for i in range(ROWS):
        y = ROWS - i - 1
        if board[(column, y)] == EMPTY:
            board[(column, y)] = color
            break
    return board


def print_color(color):
    return 'red' if color == RED else 'yellow'


def check_winning_state(board, color):
    for (x, y) in positions_in_print_order():
        for dir in [(0, 1), (1, 0), (1, 1), (-1, 1)]:
            cells = extract(board, (x, y), dir)
            if all(cell == color for cell in cells):
                log(f"Found 4-in-a-row at {x, y} dir {dir}")
                return True
    return False


def extract(board, pos, dir):
    cells = []
    for i in range(4):
        p = (pos[0] + dir[0] * i, pos[1] + dir[1] * i)
        cells.append(board[p])
    return cells


def positions_in_print_order():
    for y in range(ROWS):
        for x in range(COLUMNS):
            yield (x, y)


def convert_to_column(x):
    if x < BOARD_LEFT:
        return None
    if x > BOARD_RIGHT:
        return None
    return (x - BOARD_LEFT) // DISC_DIAMETER


def log(msg):
    pass
    # print(msg)