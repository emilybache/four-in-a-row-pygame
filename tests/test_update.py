from approvaltests import verify

from src.four_in_a_row import (GameOverState, update, GameState, ColumnWasClicked, print_color,
                               EMPTY, RED, YELLOW)


def print_model(model):
    state_string = model.__class__.__name__ + '\n'
    if isinstance(model, GameOverState):
        state_string += f'{print_color(model.winner).title()} won.\n'
    if isinstance(model, GameState):
        state_string += f'It is {model.whos_turn()}s turn.\n'
        symbols = {
            EMPTY: 'O',
            RED: 'R',
            YELLOW: 'Y'
        }
        for y in range(6):
            for x in range(7):
                pos = (x, y)
                board = model.board
                cell = board[pos]
                state_string += symbols[cell] + ' '
            state_string += '\n'
    return state_string


def test_first_placed_brick_is_red():
    model = GameState()
    model = update(model, ColumnWasClicked(0))
    state_string = print_model(model)
    verify(state_string)


def test_placing_4_bricks_in_first_column():
    model = GameState()
    model = update(model, ColumnWasClicked(0))
    model = update(model, ColumnWasClicked(0))
    model = update(model, ColumnWasClicked(0))
    model = update(model, ColumnWasClicked(0))
    state_string = print_model(model)
    verify(state_string)


def test_letting_red_win():
    model = GameState()
    model = update(model, ColumnWasClicked(0))  # red
    model = update(model, ColumnWasClicked(1))  # yellow
    model = update(model, ColumnWasClicked(0))  # red
    model = update(model, ColumnWasClicked(1))  # yellow
    model = update(model, ColumnWasClicked(0))  # red
    model = update(model, ColumnWasClicked(1))  # yellow
    model = update(model, ColumnWasClicked(0))  # red
    state_string = print_model(model)
    verify(state_string)


def test_letting_red_win_horisontally():
    model = GameState()
    model = update(model, ColumnWasClicked(0))  # red
    model = update(model, ColumnWasClicked(5))  # yellow
    model = update(model, ColumnWasClicked(1))  # red
    model = update(model, ColumnWasClicked(5))  # yellow
    model = update(model, ColumnWasClicked(2))  # red
    model = update(model, ColumnWasClicked(5))  # yellow
    model = update(model, ColumnWasClicked(3))  # red
    state_string = print_model(model)
    verify(state_string)
