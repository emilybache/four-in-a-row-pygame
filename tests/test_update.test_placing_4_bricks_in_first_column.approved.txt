GameState
It is reds turn.
O O O O O O O 
O O O O O O O 
Y O O O O O O 
R O O O O O O 
Y O O O O O O 
R O O O O O O 
